### PR TITLE
World wrap Part 2

### PR DESCRIPTION
**Pull request by GGGuenni**
[Depends on v0.5]
With this PR units can finally walk and see over the edge of the world.
It's achieved by making the tiles on the left and the right literal neighbors.

This is the last PR before v1.0

### CODE REVIEW

File: core/src/com/unciv/logic/HexMath.kt [hunk 1]
```diff
@@ -136,21 +136,32 @@ object HexMath {
...
+                    vectors += origin.cpy().scl(2f).sub(current) // Get vector on other side of clock
+            } else {
+                vectors += origin.cpy().scl(2f).sub(current) // Get vector on other side of clock
+            }
             current.add(1f, 1f)
         }
         for (i in 0 until distance) { // 10 to 12
             vectors += current.cpy()
-            vectors += origin.cpy().scl(2f).sub(current) // Get vector on other side of clock
+            if (worldWrap){
```
Comment (yairm210) [inline core/src/com/unciv/logic/HexMath.kt:149]: If(!worldWrap)...
Else if()...

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 2]
```diff
@@ -129,6 +127,30 @@ class TileMap {
                     }
                 }.filterNotNull()
 
+    private fun getIfTileExistsOrNull(x: Int, y: Int) : TileInfo? {
+        if (contains(x, y)) {
+            return get(x, y)
+        }
+
+        if (mapParameters.worldWrap) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/TileMap.kt:135]: If(worldWrap) return null
...


To reduce indentation and increase readability

File: core/src/com/unciv/logic/HexMath.kt (hunk 1 above)
Comment (GGGuenni) [reply core/src/com/unciv/logic/HexMath.kt]: That's a v0.4 feature
Reduced the size and indentation and propagated it to this version

File: core/src/com/unciv/logic/map/TileMap.kt (hunk 2 above)
Comment (GGGuenni) [reply core/src/com/unciv/logic/map/TileMap.kt]: done

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 3]
```diff
@@ -44,15 +44,15 @@ class TileMap {
...
+        for (vector in HexMath.getVectorsInDistance(Vector2.Zero, radius, worldWrap))
             tileList.add(TileInfo().apply { position = vector; baseTerrain = Constants.grassland })
         setTransients(ruleset)
     }
 
     /** generates a rectangular map of given width and height*/
-    constructor(width: Int, height: Int, ruleset: Ruleset) {
-        for (x in -width / 2..width / 2)
+    constructor(width: Int, height: Int, ruleset: Ruleset, worldWrap: Boolean = false) {
+        for (x in -width / 2 .. if (worldWrap) { width / 2 - 1 } else { width / 2 })
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/TileMap.kt:55]: Put this in a val.
val halfway = if(worldWrap) width/2-1 else width/2

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 4]
```diff
@@ -129,6 +127,31 @@ class TileMap {
...
+            return get(x, y)
+
+        if (!mapParameters.worldWrap)
+            return null
+
+        var radius = mapParameters.size.radius
+        if (mapParameters.shape == MapShape.rectangular)
+            radius = HexMath.getEquivalentRectangularSize(radius).x.toInt() / 2
+
+        if (!contains(x, y)) { //tile is outside of the map
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/TileMap.kt:141]: Surely this would always be 'true' since we check contains(x,y) at the first line in the function?

**Comment by yairm210** [comment]
Looks good!

**Comment by yairm210** [comment]
Once YOU feel that you're ready with this, I'm ready to merge this in as long as we're sure that unless toggled on it doesn't affect anyone.
There are bound to be more things to deal with gradually, but I'm sure that some of the avid Discorders will swamp us with things that aren't working the second we get a prototype out :)

**Comment by GGGuenni** [comment]
> There are bound to be more things to deal with gradually, but I'm sure that some of the avid Discorders will swamp us with things that aren't working the second we get a prototype out :)

Definitely :D 

One thing before I resolve your last review:
To merge everything at once you would have to merge #3603 only (I still have to propagate the new changes to it) else you would have to merge #3592, #3604, #3605, #3606, #3609(this one), and #3603 in this order 😬 I can also close a few PR's and merge #3592, #3604, #3605, and #3606 into one PR to make it more clustered. You would have to merge #3606, #3609 (they are completely separated from each other), and #3603 then.
So how do you want to merge? Asking because you only reviewed this one PR which isn't everything the world wrap feature needs to function. Sorry for the confusion my PR's probably made 😅

**Comment by yairm210** [comment]
From what I see, the PRs override each other. So the changed files in v0.6 contain all changed files in previous versions, no?

Since everything here is toggleable, I would like us to start with a PR containing the toggle and whatever else you deem 'minimal', and from there we can continue working out the finer details of everything else

**Comment by yairm210** [comment]
That is to say, I think all previous PRs are made redundant and obsolete by the subsequent ones

**Comment by GGGuenni** [comment]
Nope, all versions below and including 0.3.5 (#3606) are not included in this PR these two are entirely separated.
1.0 (#3603) is containing everything

**Comment by yairm210** [comment]
This also looks good!
We'll merge this after part 1 :D

**Pull request merged**